service/policies.go

```diff
@@ -0,0 +1,71 @@
+package service
+
+import (
+	"reflect"
+
+	v1 "code.vegaprotocol.io/protos/vega/wallet/v1"
+	"github.com/golang/protobuf/jsonpb"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+type ConsentConfirmation struct {
+	TxStr    string
+	Decision bool
+}
+
+type ConsentRequest struct {
+	tx *v1.SubmitTransactionRequest
+}
+
+func (r *ConsentRequest) String() (string, error) {
+	data, err := protojson.Marshal(r.tx)
+	return string(data), err
+}
+
+type Policy interface {
+	Ask(tx *v1.SubmitTransactionRequest) bool
+}
+
+type AutomaticConsentPolicy struct {
+	pendingEvents chan ConsentRequest
+	confirmations chan ConsentConfirmation
+}
+
+func NewAutomaticConsentPolicy(pending chan ConsentRequest, response chan ConsentConfirmation) Policy {
+	return &AutomaticConsentPolicy{
+		pendingEvents: pending,
+		confirmations: response,
+	}
+}
+
+func (p *AutomaticConsentPolicy) Ask(tx *v1.SubmitTransactionRequest) bool {
+	return true
+}
+
+type ExplicitConsentPolicy struct {
+	pendingEvents chan ConsentRequest
+	confirmations chan ConsentConfirmation
+}
+
+func NewExplicitConsentPolicy(pending chan ConsentRequest, response chan ConsentConfirmation) Policy {
+	return &ExplicitConsentPolicy{
+		pendingEvents: pending,
+		confirmations: response,
+	}
+}
+
+func (p *ExplicitConsentPolicy) Ask(tx *v1.SubmitTransactionRequest) bool {
+	p.pendingEvents <- ConsentRequest{tx}
+
+	for c := range p.confirmations {
```

Comment:
If I understand correctly this channel that we look at from for confirmation can have multiple readers if, for example, two transactions come in at the same time. What I find concerning in this code is that if we find an answer for a transaction that doesn't match we just ignore it. So either one of the two things are true:

1) responses will get lost. For example, if `req1` asks the listens, then `req2` asks and listens, and the response for `req1` is sent to the listener for `req2` is skips over it and `req1` never receives its reply.

2) they repsonses always come through the channels to the listeners in the order they were sent, in which case why are we doing any of this matching code?

Do we know which it is? (or if there is a third possibility please tell!)

Reply:
Basically what we have in mind with Valentin is to cut down in two PRs, this one takes care of the basis oif approval and the next one adds a queue with memory to avoid trashing requests that got there too early or too late.

Reply:
I'm even more concerned if there are other know short-falls of this implementation. Where is the ticket for the second half of the work? What are the other known issues?

I feel like all the issues can be fixed just by sending a response channel alongside the confirmation request, so that the decision can be sent directly to the right request.
```
type ConsentRequest struct {
        retChan  chan bool
	tx            *v1.SubmitTransactionRequest
}

// then after asking for user input the main thread just needs to do
go func(){
    req.retChan <- decision
    close(req.retChan)
}()
```